DataFixerUpper module

Leo40Git opened the pull request:
This PR adds the `datafixerupper` module (and its enclosing `misc` library).

Original code is by i5, and has been updated to 1.19 by me. Here's proof i5 consented to my evil code hackery:
![100% legit](https://user-images.githubusercontent.com/7997354/178154631-27084407-cbb0-4959-a613-de4ca33b5640.png)
(this code was originally from a Fabric API PR that was never merged, so I didn't include the usual "FabricMC" mention in the license.)

Closes #7.

Leo40Git committed:
Add DataFixerUppper module

Leo40Git committed:
QuiltDataFixesInternals: simplify getModSchema()

Add empty testmod to fix build

Leo40Git committed:
Add QuiltDataFixerBuilder

williambl commented:
Oooh, I like this. Lots of us don't make datafixers because of perceived difficulty, this will help lots.

Is it possible to have two test mods? One which writes some old data, one which uses a datafixer on it?

Leo40Git commented:
> One which writes some old data, one which uses a datafixer on it?

...hm, that's an idea.
I'll see what I can do tomorrow-ish.

williambl commented:
Great :D

Also, there's an extra 'p' in the title of the PR :P

Leo40Git commented:
n-no there isn't

Leo40Git committed:
begin work on testmod

Leo40Git committed:
licemse

Leo40Git committed:
Merge branch '1.19' of https://github.com/QuiltMC/quilt-standard-libraries into feature/dfu

EnnuiL commented on library/misc/datafixerupper/src/main/java/org/quiltmc/qsl/datafixerupper/api/QuiltDataFixerBuilder.java:
```diff
@@ -0,0 +1,50 @@
+	 */
+	public DataFixer build(Supplier<Executor> executorGetter) {
+		return switch (SharedConstants.DATA_FIXER_MODE) {
+		case UNINITIALIZED_UNOPTIMIZED, INITIALIZED_UNOPTIMIZED -> buildUnoptimized();
```
Nitpick: Add a tab before both `cases`

EnnuiL commented on library/misc/datafixerupper/src/main/java/org/quiltmc/qsl/datafixerupper/api/QuiltDataFixes.java:
```diff
@@ -0,0 +1,101 @@
+		checkNotNull(modId, "modId cannot be null");
+		QuiltDataFixesInternals.DataFixerEntry entry = QuiltDataFixesInternals.getFixerEntry(modId);
+		if (entry == null)
+			return Optional.empty();
```
this is cursed; Remove the line break

EnnuiL reviewed (approved):
I'm happy with this PR! I only have really tiny nitpicks, but yeah, this doesn't block an approval or anything

Leo40Git committed:
Merge remote-tracking branch 'upstream/1.19' into feature/dfu

Leo40Git committed:
Address Ennui's review

OroArmor reviewed (approved):
Looks good. I do agree a testmod for this is hard to make since it requires multiple versions and thats not easy to make in this

Leo40Git committed:
Add testmod v2

yes, this is how we're gonna test this.

Leo40Git committed:
hhhh

Leo40Git committed:
s

Leo40Git committed:
Okay, I THINK this will work?

Leo40Git committed:
cleanupDfuTestmod -> prepareDfuTestmod

also agrees to the EULA now

Leo40Git committed:
this is a travesty.

Leo40Git committed:
Make DFU test stuff completely separate from main test stuff

Leo40Git committed:
Move testmod to testmodV1

runAutoTestServer now does NOTHING to test DFU stuff

Leo40Git committed:
Fix EULA agreement

Leo40Git committed:
Testmod complete! I know it works because it found an issue LMAO

QuiltDataFixes: rename MOD_SCHEMA to BASE_SCHEMA
QuiltDataFixes: fix BASE_SCHEMA using an old Vanilla schema as its parent instead of the *current* Vanilla schema

Leo40Git commented:
HEY LOOK I did the testmod
Anyway, let me just

LambdAurora commented on library/misc/datafixerupper/src/main/java/org/quiltmc/qsl/datafixerupper/api/QuiltDataFixes.java:
```diff
@@ -0,0 +1,108 @@
+	/**
+	 * Checks if the data fixer registry is frozen.
+	 *
+	 * @return {@code true} if frozen, {@code false} otherwise.
```
```suggestion
	 * @return {@code true} if frozen, or {@code false} otherwise.
```

LambdAurora commented on library/misc/datafixerupper/src/main/java/org/quiltmc/qsl/datafixerupper/api/QuiltDataFixes.java:
```diff
@@ -0,0 +1,108 @@
+	 * Gets a mod's data version from a {@link NbtCompound}.
+	 *
+	 * @param compound the compound
+	 * @param modId the mod ID
```
```suggestion
	 * @param modId the mod identifier
```

LambdAurora commented on library/misc/datafixerupper/src/main/java/org/quiltmc/qsl/datafixerupper/api/QuiltDataFixes.java:
```diff
@@ -0,0 +1,108 @@
+	 *
+	 * @param compound the compound
+	 * @param modId the mod ID
+	 * @return the mod's data version, or 0 if the compound has no data for that mod
```
```suggestion
	 * @return the mod's data version, or {@code 0} if the compound has no data for that mod
```

LambdAurora commented on library/misc/datafixerupper/src/main/java/org/quiltmc/qsl/datafixerupper/api/QuiltDataFixes.java:
```diff
@@ -0,0 +1,108 @@
+	/**
+	 * Gets a mod's data fixer.
+	 *
+	 * @param modId the mod ID
```
```suggestion
	 * @param modId the mod identifier
```

LambdAurora commented on library/misc/datafixerupper/src/main/java/org/quiltmc/qsl/datafixerupper/api/QuiltDataFixes.java:
```diff
@@ -0,0 +1,108 @@
+	/**
+	 * Registers a new data fixer.
+	 *
+	 * @param modId the ID of the mod
```
```suggestion
	 * @param modId the mod identifier
```

LambdAurora commented on library/misc/datafixerupper/src/main/java/org/quiltmc/qsl/datafixerupper/api/QuiltDataFixes.java:
```diff
@@ -0,0 +1,108 @@
+	private QuiltDataFixes() { }
+
+	/**
+	 * A "base" version 0 schema, for use by all mods.<br>
```
```suggestion
	 * A "base" version {@code 0} schema, for use by all mods.
	 * <p>
```

LambdAurora commented on library/misc/datafixerupper/src/main/java/org/quiltmc/qsl/datafixerupper/api/SimpleFixes.java:
```diff
@@ -0,0 +1,96 @@
+	 *
+	 * @param builder the builder
+	 * @param name the fix's name
+	 * @param oldId the block's old ID
+	 * @param newId the block's new ID
```
```suggestion
	 * @param oldId the block's old identifier
	 * @param newId the block's new identifier
```

LambdAurora commented on library/misc/datafixerupper/src/main/java/org/quiltmc/qsl/datafixerupper/api/SimpleFixes.java:
```diff
@@ -0,0 +1,96 @@
+	 *
+	 * @param builder the builder
+	 * @param name the fix's name
+	 * @param oldId the item's old ID
+	 * @param newId the item's new ID
```
```suggestion
	 * @param oldId the item's old identifier
	 * @param newId the item's new identifier
```

LambdAurora commented on library/misc/datafixerupper/src/main/java/org/quiltmc/qsl/datafixerupper/api/SimpleFixes.java:
```diff
@@ -0,0 +1,96 @@
+	 *
+	 * @param builder the builder
+	 * @param name the fix's name
+	 * @param changes a map of old biome IDs to new biome IDs
```
```suggestion
	 * @param changes a map of old biome identifiers to new biome identifiers
```

LambdAurora commented on library/misc/datafixerupper/src/main/java/org/quiltmc/qsl/datafixerupper/mixin/NbtHelperMixin.java:
```diff
@@ -0,0 +1,39 @@
+
+@Mixin(NbtHelper.class)
+public abstract class NbtHelperMixin {
+	@Inject(method = "update(Lcom/mojang/datafixers/DataFixer;Lnet/minecraft/datafixer/DataFixTypes;Lnet/minecraft/nbt/NbtCompound;II)Lnet/minecraft/nbt/NbtCompound;",
+			at = @At("RETURN"), cancellable = true)
```
```suggestion
	@Inject(
			method = "update(Lcom/mojang/datafixers/DataFixer;Lnet/minecraft/datafixer/DataFixTypes;Lnet/minecraft/nbt/NbtCompound;II)Lnet/minecraft/nbt/NbtCompound;",
			at = @At("RETURN"),
			cancellable = true
	)
```

LambdAurora commented on library/misc/datafixerupper/src/main/java/org/quiltmc/qsl/datafixerupper/mixin/client/HotbarStorageMixin.java:
```diff
@@ -0,0 +1,38 @@
+
+@Mixin(HotbarStorage.class)
+public abstract class HotbarStorageMixin {
+	@Inject(method = "save",
+			at = @At(value = "INVOKE", target = "Lnet/minecraft/nbt/NbtIo;write(Lnet/minecraft/nbt/NbtCompound;Ljava/io/File;)V"),
+			locals = LocalCapture.CAPTURE_FAILHARD)
```
```suggestion
	@Inject(
			method = "save",
			at = @At(value = "INVOKE", target = "Lnet/minecraft/nbt/NbtIo;write(Lnet/minecraft/nbt/NbtCompound;Ljava/io/File;)V"),
			locals = LocalCapture.CAPTURE_FAILHARD
	)
```

LambdAurora commented on library/misc/datafixerupper/src/main/java/org/quiltmc/qsl/datafixerupper/api/QuiltDataFixes.java:
```diff
@@ -0,0 +1,108 @@
+	 * @param currentVersion the current version of the mod's data
+	 * @param dataFixer the data fixer
+	 */
+	public static void registerFixer(String modId, int currentVersion, DataFixer dataFixer) {
```
Wondering if it would be worth it to include overloads using the ModContainer class, so it creates an incentive to re-use the mod container given in the ModInitializer.

LambdAurora commented on library/misc/datafixerupper/src/main/java/org/quiltmc/qsl/datafixerupper/api/SimpleFixes.java:
```diff
@@ -0,0 +1,96 @@
+	private SimpleFixes() { }
+
+	/**
+	 * Adds a block rename fix to the builder.
```
Definitely requires an explanation of what the identifiers are meant to be

Are they supposed to respect the normal identifiers format?
Why a string?

Leo40Git replied:
Data fixers use raw strings, since they're not tied to Minecraft (DFU is a separate library).

AFAIK, these strings are supposed to respect the identifier format. I will add this to the javadoc.

Leo40Git replied:
Actually, I have a better idea....

LambdAurora reviewed (changes requested):
Lot of documentation changes :p

I also would really like a package-info explaining how you're supposed to use all of these.

Leo40Git committed:
Apply suggestions from code review

Co-authored-by: LambdAurora <aurora42lambda@gmail.com>

Leo40Git committed:
SimpleFixes: make methods accept identifiers as Identifiers

Leo40Git committed:
QuiltDataFixes: add ModContainer overload to registerFixer

Leo40Git committed:
Fix up testmods

Leo40Git committed:
Create package-info.java

Leo40Git committed:
Format param javadocs

Leo40Git committed:
Prevent data fixers from being replaced

Leo40Git committed:
Merge branch '1.19' into feature/dfu

LambdAurora commented on library/misc/datafixerupper/src/main/java/org/quiltmc/qsl/datafixerupper/api/package-info.java:
```diff
@@ -0,0 +1,38 @@
+ *     Schema schemaV1 = builder.addSchema(1, IdentifierNormalizingSchema::new)
+ *     // add fixes to the schema - for example, an item rename (identifier change)
+ *     SimpleFixes.addItemRenameFix(builder, "Rename cool_item to awesome_item",
+ *         new Identifier("mymod", "cool_item", new Identifier("mymod", "awesome_item", schemaV1);
```
```suggestion
 *         new Identifier("mymod", "cool_item"), new Identifier("mymod", "awesome_item"), schemaV1);
```

LambdAurora commented on library/misc/datafixerupper/src/main/java/org/quiltmc/qsl/datafixerupper/mixin/ChunkSerializerMixin.java:
```diff
@@ -0,0 +1,35 @@
+
+@Mixin(ChunkSerializer.class)
+public abstract class ChunkSerializerMixin {
+	@ModifyVariable(method = "serialize",
+			at = @At(value = "INVOKE", target = "Lnet/minecraft/nbt/NbtCompound;putInt(Ljava/lang/String;I)V", ordinal = 0))
```
```suggestion
	@ModifyVariable(
			method = "serialize",
			at = @At(value = "INVOKE", target = "Lnet/minecraft/nbt/NbtCompound;putInt(Ljava/lang/String;I)V", ordinal = 0)
	)
```

LambdAurora commented on library/misc/datafixerupper/src/main/java/org/quiltmc/qsl/datafixerupper/api/QuiltDataFixes.java:
```diff
@@ -0,0 +1,125 @@
+	 * @param currentVersion the current version of the mod's data
+	 * @param dataFixer      the data fixer
+	 */
+	public static void registerFixer(ModContainer mod, int currentVersion, DataFixer dataFixer) {
```
Ok, sorry for not mentioning it earlier, but new day new thoughts

What if the current version was something declared in the QMJ so it would be easy to locate?
Would keep this overload still, but it's an interesting thought to have.

Leo40Git replied:
So something like `"data_version": int`? Hm...

Leo40Git replied:
Er, wait, problem: the `DataFixerBuilder` constructor takes the latest data version... meaning if a user declares it in their `quilt.mod.json`, they'd need to read it themselves anyway.

LambdAurora replied:
This method has the ModContainer tho?

Leo40Git replied:
```java
public static void initialize(ModContainer mod) {
    // END-USER CODE that needs the data version
    QuiltDataFixerBuilder builder = new QuiltDataFixerBuilder(DATA_VERSION);
    builder.addSchema(0, QuiltDataFixes.BASE_SCHEMA);
    // etc, etc...
    
    // LIBRARY CALL that needs the data version
    QuiltDataFixes.registerFixer(mod, DATA_VERSION, builder.build(Util::getBootstrapExecutor));
}
```
Making the LIBRARY CALL not need the data version is pointless, since the END-USER CODE still needs it.

LambdAurora replied:
ooooh

Now we see.

Leo40Git committed:
Apply suggestions from code review

Co-authored-by: LambdAurora <aurora42lambda@gmail.com>

Leo40Git committed:
package-info.java: slight reformatting

Leo40Git committed:
Add NotNull annotations

LambdAurora commented on library/misc/datafixerupper/src/main/java/org/quiltmc/qsl/datafixerupper/api/QuiltDataFixes.java:
```diff
@@ -0,0 +1,126 @@
+	 *
+	 * @return {@code true} if frozen, or {@code false} otherwise.
+	 */
+	public static boolean isFrozen() {
```
```suggestion
	@Contract(pure = true)
	public static boolean isFrozen() {
```

Would also need to annotate the internal one to make sure the contract is always respected.

LambdAurora reviewed (commented):
Since you went with annotating stuff, here's some feedback on annotating:

- I suggested a contract one for isFrozen
- we might want to add the same contract to getModDataVersion
- I believe versions can only be between `0` and Integer.MAX_INTEGER, if that's the case then they should be annotated with `@Range`

Leo40Git committed:
QuiltDataFixes: add contract to isFrozen

Co-authored-by: LambdAurora <aurora42lambda@gmail.com>

Leo40Git commented:
Oh, I didn't realize `@Range` existed!

Leo40Git committed:
QuiltDataFixes: more annotation work

Leo40Git committed:
QuiltDataFixes: mark getModDataVersion as pure

QuiltDataFixesInternals: mirror API annotations

LambdAurora commented on library/misc/datafixerupper/src/main/java/org/quiltmc/qsl/datafixerupper/api/QuiltDataFixes.java:
```diff
@@ -0,0 +1,135 @@
+	 * @return the mod's data version, or {@code 0} if the compound has no data for that mod
+	 */
+	@Contract(pure = true)
+	public static int getModDataVersion(@NotNull NbtCompound compound, @NotNull String modId) {
```
```suggestion
	public static @Range(from = 0, to = Integer.MAX_VALUE) int getModDataVersion(@NotNull NbtCompound compound, @NotNull String modId) {
```

LambdAurora commented on library/misc/datafixerupper/src/main/java/org/quiltmc/qsl/datafixerupper/api/QuiltDataFixerBuilder.java:
```diff
@@ -0,0 +1,51 @@
+ * An extended variant of the {@link DataFixerBuilder} class, which provides an extra method.
+ */
+public class QuiltDataFixerBuilder extends DataFixerBuilder {
+	public QuiltDataFixerBuilder(int dataVersion) {
```
```suggestion
	public QuiltDataFixerBuilder(@Range(from = 0, to = Integer.MAX_VALUE) int dataVersion) {
```

LambdAurora reviewed (commented):
After that I'll approve :P

Leo40Git committed:
QuiltDataFixerBuilder.build: contractttttttt

Leo40Git committed:
QuiltDataFixes[Internals]: add Range to getModDataVersion

Leo40Git committed:
QuiltDataFixerBuilder: add Range to version in constructor

Ellemes reviewed (approved):
LGTM, tested with my mod and it seems to work fine.

Leo40Git committed:
QuiltDataFixes: add registerFixer(ModContainer, QuiltDataFixerBuilder)

Leo40Git committed:
testmodV1: don't register (empty) data fixer

Leo40Git committed:
testmodV2: add block, prepare for testmodV3

Leo40Git committed:
testmodV1, testmodV2: write marker files in game dir

Leo40Git committed:
QuiltDataFixes: rename registerFixer(ModContainer, QuiltDataFixerBuilder) to buildAndRegisterFixer

Leo40Git committed:
Add testmod v3

Leo40Git committed:
testmodV3: fix new_item to fine_item rename

Leo40Git committed:
Merge remote-tracking branch 'upstream/1.19' into feature/dfu

Leo40Git committed:
Revert to testmod v2

Leo40Git committed:
Disable apparently-broken testmods for now

Leo40Git commented:
SO APPARENTLY I BROKE MY TESTMODS (edit: running on CI, locally they work fine)
Well they were always kind of a kludge but now it's failing for a reason I can't even decipher

LambdAurora committed:
Merge branch '1.19' into feature/dfu

LambdAurora committed:
Merge branch '1.19' into feature/dfu

LambdAurora committed:
Attempt to run DFU testmods in CI.

LambdAurora committed:
Gradle is pain, fix more stuff so everything plays nicely.
